File: frameworks/java/javalin/Implementation/src/main/java/implementation/Serial.java
package implementation;

import java.util.concurrent.atomic.AtomicLong;

import io.javalin.http.Context;

/**
 * x-rb-serial: one counter for the whole process. A handler that writes it increments it and
 * writes the new value, so an answer the cache replays carries the value it was stored with.
 */
public final class Serial {

    public static final String HEADER = "x-rb-serial";

    private static final AtomicLong LAST = new AtomicLong();

    private Serial() {}

    public static void write(Context ctx) {
        ctx.header(HEADER, Long.toString(LAST.incrementAndGet()));
    }
}

File: frameworks/rust/rocket/Implementation/serial.rs
use std::sync::atomic::{AtomicU64, Ordering};

use rocket::Responder;
use rocket::http::Header;

static LAST: AtomicU64 = AtomicU64::new(0);

/// An answer with x-rb-serial: one counter for the whole process. A handler that writes it takes
/// the next value, so an answer a cache replays carries the value it was stored with.
#[derive(Responder)]
pub struct Fresh<R> {
    inner: R,
    serial: Header<'static>,
}

impl<R> Fresh<R> {
    pub fn new(inner: R) -> Self {
        Fresh { inner, serial: next() }
    }
}

pub fn next() -> Header<'static> {
    Header::new("x-rb-serial", (LAST.fetch_add(1, Ordering::Relaxed) + 1).to_string())
}
